app_dart/lib/src/request_handlers/get_timeseries_history.dart

```diff
@@ -0,0 +1,117 @@
+  const GetTimeSeriesHistory(
+    Config config, {
+    @visibleForTesting DatastoreServiceProvider datastoreProvider,
+  })  : datastoreProvider =
+            datastoreProvider ?? DatastoreService.defaultProvider,
+        super(config: config);
+
+  final DatastoreServiceProvider datastoreProvider;
+  static const String timeSeriesKeyParam = 'TimeSeriesKey';
+  static const String startFromParam = 'StartFrom';
+
+  @override
+  Future<GetTimeSeriesHistoryResponse> post() async {
+    checkRequiredParameters(<String>[timeSeriesKeyParam]);
+    const int maxRecords = 1500;
+    final DatastoreService datastore = datastoreProvider();
+    final KeyHelper keyHelper = KeyHelper(applicationContext: AppEngineContext(false,'','','','','',Uri()));
+    final Set<Commit> commits =
+        await datastore.queryRecentCommits(limit: maxRecords).toSet();
+
+    Key timeSeriesKey;
+    try {
+      timeSeriesKey = keyHelper.decode(requestData[timeSeriesKeyParam]);
+    } catch (error) {
+      throw BadRequestException(
+          'Bad timeSeries key: ${requestData[timeSeriesKeyParam]}');
+    }
+
+    final TimeSeries timeSeries =
+        await config.db.lookupValue<TimeSeries>(timeSeriesKey, orElse: () {
+      throw BadRequestException('No such timeseries: ${timeSeriesKey.id}');
+    });
+
+    List<TimeSeriesValue> timeSeriesValues = await datastore
+        .queryRecentTimeSeriesValues(timeSeries,
+            startFrom: startFromParam, limit: maxRecords)
+        .toList();
+    timeSeriesValues = insertMissingTimeseriesValues(timeSeriesValues, commits);
+
+    return GetTimeSeriesHistoryResponse(timeSeries, timeSeriesValues);
+  }
+}
+
+@immutable
+class GetTimeSeriesHistoryResponse extends JsonBody {
+  const GetTimeSeriesHistoryResponse(this.timeSeries, this.timeSeriesValues)
+      : assert(timeSeries != null);
+
+  final TimeSeries timeSeries;
+  final List<TimeSeriesValue> timeSeriesValues;
+
+  @override
+  Map<String, dynamic> toJson() {
+    return <String, dynamic>{
+      'BenchmarkData': BenchmarkData(
+          timeSeriesEntity: TimeseriesEntity(timeSeries: timeSeries),
+          values: timeSeriesValues),
+      //LastPosition to do @Keyong
+      'LastPosition': null,
+    };
+  }
+}
+
+List<TimeSeriesValue> insertMissingTimeseriesValues(
+    List<TimeSeriesValue> timeSerialsValues, Set<Commit> commits) {
+  final List<TimeSeriesValue> values = <TimeSeriesValue>[];
+  final Map<String, TimeSeriesValue> valuesByCommit =
+      <String, TimeSeriesValue>{};
+  for (TimeSeriesValue value in timeSerialsValues) {
+    valuesByCommit[value.revision] = value;
+  }
+
+  for (Commit commit in commits) {
+    TimeSeriesValue value;
+
+    if (valuesByCommit.containsKey(commit.sha)) {
+      value = valuesByCommit[commit.sha];
+      if (value.value < 0) {
+        value.value = 0;
+      }
+    } else {
+      // Insert placeholder entries for missing values.
```

Comment:
When do we expect missing values?  Maybe that could be added to the comment here.

Reply:
Added comments like:
      // Insert placeholder entries for missing values. 
      // Missing values happen when there is any queried timeSeriesValue 
      // whose commit does not appear in the queried top [maxRecords] commits

Reply:
This may due to [issue 420](https://github.com/flutter/cocoon/issues/420).